posthog/hogql_queries/sessions_timeline_query_runner.py

```diff
@@ -0,0 +1,191 @@
+                    right=ast.Call(
+                        name="toDateTime",
+                        args=[ast.Constant(value=self.query.after)],
+                    ),
+                ),
+                ast.CompareOperation(
+                    op=ast.CompareOperationOp.Lt,
+                    left=ast.Field(chain=["timestamp"]),
+                    right=ast.Call(
+                        name="toDateTime",
+                        args=[ast.Constant(value=self.query.before)],
+                    ),
+                ),
+            ]
+            if self.query.personId:
+                event_conditions.append(
+                    ast.CompareOperation(
+                        left=ast.Field(chain=["person_id"]),
+                        right=ast.Constant(value=self.query.personId),
+                        op=ast.CompareOperationOp.Eq,
+                    )
+                )
+            select_query = cast(
+                ast.SelectQuery,
+                parse_select(
+                    """
+                    SELECT
+                        e.uuid,
+                        e.timestamp,
+                        e.event,
+                        e.properties,
+                        e.distinct_id,
+                        e.elements_chain,
+                        e.session_id AS formal_session_id,
+                        first_value(e.uuid) OVER (
+                            PARTITION BY (e.person_id, session_id_flip_index) ORDER BY __toInt64(timestamp)
+                            RANGE BETWEEN 1800 PRECEDING AND CURRENT ROW /* split informal session after 30+ min */
+                        ) AS informal_session_uuid,
+                        dateDiff('s', sre.start_time, sre.end_time) AS recording_duration_s
+                    FROM (
+                        SELECT
+                            *,
+                            sum(session_id = prev_session_id ? 0 : 1) OVER (
+                                PARTITION BY person_id ORDER BY timestamp ROWS UNBOUNDED PRECEDING
+                            ) AS session_id_flip_index
+                        FROM (
+                            SELECT
+                                uuid,
+                                person_id AS person_id,
+                                timestamp AS timestamp,
+                                event,
+                                properties,
+                                distinct_id,
+                                elements_chain,
+                                $session_id AS session_id,
+                                lagInFrame($session_id, 1) OVER (
+                                    PARTITION BY person_id ORDER BY timestamp
+                                ) AS prev_session_id
+                            FROM events
+                            WHERE {event_conditions}
+                            ORDER BY timestamp DESC
+                            LIMIT {event_limit_with_more}
```

Comment:
How would we paginate this? Looking for the lowest timestamp in the response and then rerunning with this timestamp as `before`?

Reply:
Exactly